portfolio/src/main/java/com/google/sps/data/TimeRangeGroupArrayList.java

```diff
@@ -0,0 +1,121 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+/** Models an implementation of the TimeRangeGroup model using ArrayList. */
+public class TimeRangeGroupArrayList implements TimeRangeGroup {
+
+  public List<TimeRange> allTimeRanges;
+  private Comparator<TimeRange> comparator;
+  private boolean ascending;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public TimeRangeGroupArrayList(
+      Collection<TimeRange> timeRanges, Comparator<TimeRange> comparator, boolean ascending) {
+    allTimeRanges = new ArrayList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+    this.comparator = comparator;
+    this.ascending = ascending;
+    sortTimeRanges(comparator, ascending);
+  }
+
+  /**
+   * Adds a new time range to the list. This new range must be disjoint from all existing ranges.
+   * TODO(hollyyuqizheng): Depending on how the algorithms are implemented, we might need to add the
+   * merging logic in this method. For example, if a time range is deleted but then later added back
+   * to the list for some reason, this list will potentially need to be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    for (TimeRange currentRange : allTimeRanges) {
+      if (currentRange.overlaps(timeRange)) {
+        throw new IllegalArgumentException(
```

Comment:
The throwing of this exception seems a little odd to me, and I'm having a hard time seeing why we'd want to do this. Is there any way for the caller to know that they're about to trigger this? The `hasTimeRange(...)` method below checks if the given range is completely contained by some `TimeRange`, but doesn't tell you about overlaps. Is the caller meant to do this merging themselves somehow?

Reply:
I think ideally this class handles the overlap by merging the time range. For example, if 1pm-2pm was an existing time range, and this method was called with 1:30pm-2:30pm, then maybe this class can merge them and store 1pm-2:30pm.

But I think that merging logic does not need to happen in this CL. So maybe for now you can require that this method is called with a time range that does not overlap with any existing time range. Maybe you can throw UnsupportedOperationException and add a TODO comment.

Reply:
I agree that the class should handle the merging, but I'm not sure if it's a good idea to leave it to a different CL.

> So maybe for now you can require that this method is called with a time range that does not overlap with any existing time range.

I don't think there's a good way for the caller to know whether or not this is true is there? I guess they can call `getAllTimeRanges()` and check them all first, but that seems pretty awkward.